src/components/include/cbuf.h

```diff
@@ -235,83 +231,94 @@ cbuf_cons(u32_t cbid)
 	assert(cm->cbid_tag.cbid == id);
-	if (unlikely(CBUF_OWNER(cm))) {printc("owner\n"); goto done;}
-	if (unlikely((len >> PAGE_ORDER) > cm->sz)) {printc("too big\n"); goto done;}
-	assert(CBUF_REFCNT(cm) < CBUF_REFCNT_MAX);
+	if (unlikely(CBUF_OWNER(cm))) goto done;
+	if (unlikely((len >> PAGE_ORDER) > cm->sz)) goto done;
+	assert(!CBUF_TMEM(cm));
+
+	/* The order here is important. But as x86 does reorder stores, we only 
+	 * need compiler barrier. If the nrecv counter is increased first, the 
+	 * manager will find nsent = nreced, and both components' refcnt is 0. 
+	 * Thus it may remove this cbuf. In addition atomic instruction is 
+	 * also needed, because multiple threads can concurrently execute this code*/
 	CBUF_REFCNT_ATOMIC_INC(cm);
-	assert(cm->snd_rcv.nrecvd < CBUF_SENDRECV_MAX);
-	CBUF_NRCV_INC(cm);      //need to be atomic???
-	ret = ((void*)CBUF_PTR(cm));
+	cos_compiler_barrier();
+	CBUF_NRCV_ATOMIC_INC(cm);
+
+	/* TODO: Issue #120. As we use faa to increase the counter, check overflow 
+	 * before increment is useless, but after increment the overflow is happen. */
+	assert(CBUF_REFCNT(cm) > (unsigned int)(CBUF_REFCNT(cm)-1));
+	assert(cm->snd_rcv.nrecvd > (unsigned int)(cm->snd_rcv.nrecvd-1));
+	ret = (void*)CBUF_PTR(cm);
+	assert(ret);
 done:	
 	return ret;
 }
 
 /* 
- * This is only called for permanent cbufs.  This is called every time
- * we wish to send the cbuf to another component.  For each cbufp2buf,
- * it should be called one or more times (i.e. for each component it
- * sends the cbuf to.  If the component is instead done with the cbuf,
- * and does not want to send it anywhere, it should cbufp_free it
- * instead.  So to be more precise:
+ * This is called every time we wish to send the cbuf to another 
+ * component. For each cbuf2buf, it should be called one or more 
+ * times (i.e. for each component it sends the cbuf to. If the 
+ * component is instead done with the cbuf, and does not want to 
+ * send it anywhere, it should cbuf_free it instead.
+ * So to be more precise:
  *
- * A component that cbuf2bufs a cbuf can cbufp_send it one or more
+ * A component that cbuf2bufs a cbuf can cbuf_send it one or more
  * times to other components.  The last time we send the cbuf, free=1,
- * or the component can call cbufp_free.
+ * or the component can call cbuf_free.
  */
 static inline void
 __cbuf_send(cbuf_t cb, int free)
 {
-	u32_t id;
+	int id;
 	struct cbuf_meta *cm;
 
 	cbuf_unpack(cb, &id);
 
 	cm = cbuf_vect_lookup_addr(id);
 	assert(cm && cm->nfo);
 	assert(CBUF_REFCNT(cm));
-	assert(cm->snd_rcv.nsent < CBUF_SENDRECV_MAX);
 
-	CBUF_NSND_INC(cm);
-	if (free) CBUF_REFCNT_ATOMIC_DEC(cm);
-	/* really should deal with this case correctly */
-	assert(!CBUF_RELINQ(cm));     //???
+	int old = cm->snd_rcv.nsent;
+	CBUF_NSND_ATOMIC_INC(cm);
+	if (!(cm->snd_rcv.nsent > old)) printc("spd %d thd %d cb %d sent %d old %d\n", cos_spd_id(), cos_get_thd_id(), cb, cm->snd_rcv.nsent, old);
```

Comment:
This is hard.
I do not have good way to handle FAA overflow, because when we know the overflow, it has 
happened. Undo is not always possible.
This is why I have an assertion there.
we can discuss it if we want.